### PR TITLE
pre-commit: Replace flake8, isort, and pyupgrade with ruff

### PR DESCRIPTION
**Pull request by cclauss**
[Ruff](https://beta.ruff.rs) supports [over ~400~ 500 lint rules](https://beta.ruff.rs/docs/rules) including flake8, isort, pylint, and pyupgrade and is written in Rust for speed.
Also added rules for pylint conventions and pylint errors.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -4,22 +4,14 @@
...
-    rev: v3.2.2
-    hooks:
-      - id: pyupgrade
-        args: [--py36-plus]
-  - repo: https://github.com/PyCQA/flake8
-    rev: 5.0.4
-    hooks:
-      - id: flake8
+      - id: ruff
+        # args: [--fix, --exit-non-zero-on-fix]
```
Comment (Pierre-Sassoulas) [inline .pre-commit-config.yaml:17]: It would be convenient to have autofix, (pyupgrade was autofixing so it would be a regression not to)

**Comment by Pierre-Sassoulas** [comment]
Who can take the decision to drop python 3.6 ? It's been EOL for a long time at this point.

**Comment by RonnyPfannschmidt** [comment]
Yes we can

File: .pre-commit-config.yaml (hunk 1 above)
Comment (cclauss) [reply .pre-commit-config.yaml:17]: Can we add `--fix` in a later PR?  They tend to be more controversial and generate a lot of [redo requests](https://github.com/pytest-dev/pytest-cov/pull/583#issuecomment-1447794026).

**Comment by cclauss** [comment]
#567 drops py36

**Comment by ionelmc** [comment]
Ruff looks nice if it can replace the all the tools with a single config. Are there any downsides with ruff or anything that would need attention?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -65,7 +65,7 @@ jobs:
             toxpython: 'python3.6'
             python_arch: 'x64'
             tox_env: 'py36-pytest70-xdist250-coverage62'
-            os: 'ubuntu-latest'
+            os: 'ubuntu-20.04'
```
[the remark this answers is not in the document]
Comment (cclauss) [reply .github/workflows/test.yml]: Those lines were deleted in #567 (merged)

**Comment by cclauss** [comment]
I have not seen downsides (except maybe it is not yet v1.0) and I have been using it broadly:
https://beta.ruff.rs/docs/#testimonials

**Review by Pierre-Sassoulas — APPROVED** [review]
Apart from the fact that without fix it's not equivalent to having pyupgrade LGTM.

**Comment by ionelmc** [comment]
I've played a bit with ruff now, it's isort support is incomplete (no support for default-section or src-paths at least). The old isort hook needs to be kept.

**Comment by cclauss** [comment]
This cannot be achieved by using [forced separate and known parties](https://beta.ruff.rs/docs/settings/#forced-separate)?
* https://beta.ruff.rs/docs/settings/#isort

If not then I can drop the `I` rules from this pull request and bring back `isort`.

> [Timothy Crosley](https://twitter.com/timothycrosley/status/1606420868514877440), creator of [isort](https://github.com/PyCQA/isort):
>> Just switched my first project to Ruff. Only one downside so far: it's so fast I couldn't believe it was working till I intentionally introduced some errors.
* https://beta.ruff.rs/docs/#testimonials

File: .pre-commit-config.yaml (hunk 1 above)
Comment (ionelmc) [reply .pre-commit-config.yaml:17]: `--show-fixes` is necessary to show what has been fixed (and more importantly why).

I think enabling `--fix` in a later pr only delays the inevitable. This is like buying new alloy rims, taking the wheels off your car and throw the old rims with the old tires and then have it stuck on blocks forever cause you don't have the right tires for the new rims.

File: .ruff.toml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# [tool.ruff]  # <- TODO Uncomment when migrating to pyproject.toml
+exclude = [
+    ".eggs",
+    ".tox",
+    "build",
+    "ci/templates",
+    "dist",
+]
+line-length = 140
+select = [
```
Comment (ionelmc) [inline .ruff.toml:10]: Shouldn't this be `extend-select` instead?

Comment (cclauss) [reply .ruff.toml:10]: > If you're wondering how to configure Ruff, here are some recommended guidelines:
> * Prefer select and ignore over extend-select and extend-ignore, to make your rule set explicit.

https://beta.ruff.rs/docs/configuration

File: .ruff.toml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    "F",
+    "I",
+    "PLC",
+    "PLE",
+    "UP",
+    "W",
+]
+target-version = "py37"
+
+# [tool.ruff.isort]  # <- TODO Uncomment when migrating to pyproject.toml
```
[the remark this answers is not in the document]
Comment (cclauss) [reply .ruff.toml:21]: We can just run [`ini2toml`](https://ini2toml.readthedocs.io) to convert `setup.cfg` data into `pyproject.toml` and delete `setup.cfg` as the pip docs recommend but that is an activity of a different pull request.

File: .ruff.toml (hunk 3 above)
Comment (ionelmc) [reply .ruff.toml:10]: Well all I want is to not hardcode a ruleset (which `select = [stuff]` exactly is). I don't think any of those recommendations are worth considering blindly given the tool's age and popularity.

How about `select = ["ALL"]` and ignore whatever doesn't make sense for pytest-cov?

File: .ruff.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ionelmc) [reply .ruff.toml:21]: No.

File: .ruff.toml (hunk 3 above)
Comment (cclauss) [reply .ruff.toml:10]: > If you're wondering how to configure Ruff, here are some __recommended guidelines__:
> *    Prefer select and ignore over extend-select and extend-ignore, to make your rule set explicit.
> *    Use `ALL` with discretion. Enabling `ALL` will implicitly enable new rules whenever you upgrade.
> *    Start with a small set of rules (`select = ["E", "F"]`) and add a category at-a-time. For example, you might consider expanding to `select = ["E", "F", "B"]` to enable the popular flake8-bugbear extension.
> *    By default, Ruff's __autofix is aggressive__. If you find that it's too aggressive for your liking, consider turning off autofix for specific rules or categories (see [FAQ](https://beta.ruff.rs/docs/faq/#ruff-tried-to-fix-something--but-it-broke-my-code)).

Comment (ionelmc) [reply .ruff.toml:10]: Ok so I've played with it a bit more.... now I am not really sure what the best way... eg: these are either invalid checks or would be a pain to sort out, or not even worth fixing:
```
ignore = [
    "D", # pydocstyle
    "ANN", # flake8-annotations
    "ARG", # flake8-unused-arguments
    "BLE", # flake8-blind-except
    "C90", # mccabe
    "EM", # flake8-errmsg
    "FBT", # flake8-boolean-trap
    "INP", # flake8-no-pep420
    "PLR", # Pylint Refactor
    "PLW", # Pylint Warning
    "Q", # flake8-quotes
    "RET", # flake8-return
    "SIM102", # flake8-simplify collapsible-if
    "SIM105", # flake8-simplify use-contextlib-suppress
    "SLF", # flake8-self
    "T20", # flake8-print
    "S101", # flake8-bandit assert
    "S102", # flake8-bandit exec
    "S110", # flake8-bandit try-except-pass
    "TRY", # tryceratops
]
```

Now considering that rsync features are being dropped in xdist ... a lot of bulk using that in pytest-cov becomes unnecessary, if it isn't already as some people are suggesting pytest-cov is not necessary for getting xdist support...

I guess select with a few is fine for now.

Comment (cclauss) [reply .ruff.toml:10]: So...  How should we proceed?  Given your explorations and your knowledge of the codebase, do you want to open a new PR that cherry-picks from this one?  I have no trouble closing this with unmerged commits.  Or should I press on?

Comment (Skylion007) [reply .ruff.toml:10]: I'd also recommend the `C4` checks as they are popular, increase performance, and have autofixes enabled for all rules.

**Comment by cclauss** [comment]
As the Ruff docs say, `select=ALL` is a bad idea.  Closing.

**Pull request closed without merging**